pkg/promtail/targets/pubsub/formatter.go

```diff
@@ -0,0 +1,56 @@
+package pubsub
+
+import (
+	"encoding/json"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/prometheus/common/model"
+)
+
+// LogEntry that will be written to the pubsub topic.
+// According to the following spec.
+// https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+type GCPLogEntry struct {
+	LogName  string `json:"logName"`
+	Resource struct {
+		Type   string            `json:"type"`
+		Labels map[string]string `json:"labels"`
+	} `json:"resource"`
+	Timestamp string `json:"timestamp"`
+
+	// The time the log entry was received by Logging.
+	// Its important that `Timestamp` is optional in GCE log entry.
+	ReceiveTimestamp string `json:"receiveTimestamp"`
+
+	// TODO(kavi): Add other meta data fields as well if needed.
+}
+
+func format(m *pubsub.Message, other model.LabelSet) (api.Entry, error) {
+	var ge GCPLogEntry
+
+	if err := json.Unmarshal(m.Data, &ge); err != nil {
+		return api.Entry{}, err
+	}
+
+	labels := model.LabelSet{
+		"logName":      model.LabelValue(ge.LogName),
+		"resourceType": model.LabelValue(ge.Resource.Type),
+	}
+	for k, v := range ge.Resource.Labels {
+		labels[model.LabelName(k)] = model.LabelValue(v)
```

Comment:
MIght need to ensure these are cleaned to be Prometheus-label compatible.

Reply:
@owen-d you mean like all to `snake_case`? or you mean something else?